airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/cursor/default_file_based_cursor.py

```diff
@@ -0,0 +1,91 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import logging
+from datetime import datetime, timedelta
+from typing import Iterable, Mapping, Optional
+
+from airbyte_cdk.sources.file_based.remote_file import RemoteFile
+from airbyte_cdk.sources.file_based.stream.cursor.file_based_cursor import FileBasedCursor
+from airbyte_cdk.sources.file_based.types import StreamState
+
+DATE_TIME_FORMAT = "%Y-%m-%dT%H:%M:%S.%fZ"
+
+
+class DefaultFileBasedCursor(FileBasedCursor):
+    def __init__(self, max_history_size: Optional[int], days_to_sync_if_history_is_full: Optional[int]):
+        self._file_to_datetime_history: Mapping[str:datetime] = {}
+        self._max_history_size = max_history_size or 10_000
+        self._time_window_if_history_is_full = timedelta(days=days_to_sync_if_history_is_full or 3)
+        self._start_time = self._compute_start_time()
+
+    def set_initial_state(self, value: StreamState) -> None:
+        self._file_to_datetime_history = value.get("history", {})
+        self._start_time = self._compute_start_time()
+        self._initial_earliest_file_in_history = self._compute_earliest_file_in_history()
+
+    def add_file(self, file: RemoteFile) -> None:
+        self._file_to_datetime_history[file.uri] = file.last_modified.strftime(DATE_TIME_FORMAT)
+        if len(self._file_to_datetime_history) > self._max_history_size:
+            # Get the earliest file based on its last modified date and its uri
+            oldest_file = self._compute_earliest_file_in_history()
+            del self._file_to_datetime_history[oldest_file.uri]
+
+    def get_state(self) -> StreamState:
+        state = {
+            "history": self._file_to_datetime_history,
+        }
+        return state
+
+    def is_history_partial(self) -> bool:
+        return len(self._file_to_datetime_history) >= self._max_history_size
+
+    def _should_sync_file(self, file: RemoteFile) -> bool:
+        if file.uri in self._file_to_datetime_history:
+            # If the file's uri is in the history, we should sync the file if it has been modified since it was synced
+            return file.last_modified > datetime.strptime(self._file_to_datetime_history.get(file.uri), DATE_TIME_FORMAT)
```

Comment:
Added a warning.

I don't feel super strongly about whether we should sync the file if its last-modified is older than the one in the history, but it could have consequences on what else we sync if the history is full.

Assuming a max size of 3.

initial state:
```
history: {
  "a.csv": "2021-05-05",
  "b.csv": "2021-05-06",
  "c.csv": "2021-01-07",
}
```
Sync 1: we read new files
```
history: {
  "c.csv": "2021-01-07",
  "d.csv": "2021-05-08",
  "e.csv": "2021-05-09",
}
```

Sync 2: c.csv's last-modified is set to a value in the past (2021-05-04)
```
history: {
  "c.csv": "2021-01-04",
  "d.csv": "2021-05-08",
  "e.csv": "2021-05-09",
}
```

Sync 3: We need to sync a.csv and b.csv again because their last-modified is later than c.csv's